content/6_Advanced/Game_Theory.mdx

```diff
@@ -40,10 +38,389 @@ frequency: 0
+
+![alt text](./assets/nim.png)
+
+According to $(1)$ the nimber value is the $\text{mex}$ of the reachable nimber values, in this case it is the $\text{mex}$ of the previous $3$ values, because we can only remove 1, 2 or 3 stones in one move, so the values are computed as following:
+
+- $n(0) = 0$
+- $n(1) = \text{mex}\{0\} = 1$
+- $n(2) = \text{mex}\{0, 1\} = 2$
+- $n(3) = \text{mex}\{0, 1, 2\} = 3$
+- $n(4) = \text{mex}\{1, 2, 3\} = 0$
+- $n(5) = \text{mex}\{2, 3, 0\} = 1$
+- $n(6) = \text{mex}\{3, 0, 1\} = 2$
+- $n(7) = \text{mex}\{0, 1, 2\} = 3$
+
+A nimber values equal to $0$ represents a losing state of the game; see $n(4)$ and $n(7)$.
+
+
+Recall that a winning state means that in one move you can reach push the other player in a losing state. For example, $n(4)=0$ because we can't reach another losing state, like $n(0)$. Similarly $n(8) = 0$, because we can't reach $n(4)$ removing 1, 2 or 3 stones in one move.
+
+
+The single-pile game was a particular case, let's try extending the theorem to multiple piles of rocks. From the previous section we know that state is a losing state if the xor-sum of the piles is $0$. We can apply the same logic with the nimber values, so the nimber value of the game is equal to the xor-sum of individual games - piles- and if it's $0$, then it means the state of the game is a losing state. Mathematically this would be written as:
+
+$$
+\tag{2}
+n(\sum{p_i})=\oplus n(p_i)
+$$
+
+###
+
+### Applications
+
+#### Snim
+
+<FocusProblem problem="snim" />
+
+This problem is a direct application of $(2)$. Compute the nimber values for single piles and combine them using xor-sum.
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+#include <algorithm>
+#include <iostream>
+#include <vector>
+
+using namespace std;
+
+bool solve(const vector<int> &v, const vector<int> &vals) {
+	vector<int> nimber(*max_element(v.begin(), v.end()) + 1);
+	for (int i = 0; i < (int)nimber.size(); i++) {
+		vector<bool> seen(i + 1, false);
+		for (int val : vals) {
+			if (i >= val) { seen[nimber[i - val]] = true; }
+		}
+		int cnt = 0;
+		while (seen[cnt]) { cnt++; }
+		nimber[i] = cnt;
+	}
+	int s = 0;
+	for (int x : v) { s ^= nimber[x]; }
+	return s == 0;
+}
+
+int main() {
+	int k;
+	cin >> k;
+	vector<int> a(k);
+	for (int &val : a) { cin >> val; }
+	int t;
+	cin >> t;
+	while (t--) {
+		int n;
+		cin >> n;
+		vector<int> v(n);
+		for (int &val : v) { cin >> val; }
+		cout << (solve(v, a) ? 'L' : 'W');
+	}
+	return 0;
+}
+```
+
+</CPPSection>
+</LanguageSection>
+
+#### Strange Nim
+
+<FocusProblem problem="strange-nim" />
+
+We can apply the Sprague-Grundy theorem $(2)$. Let $g(n,k)$ be the nimber value for a particular pile. To speed up the algorithm, a few observations are necessary:
+1. $g(n, k) = n/k \iff k \mid n$
+2. $g(n, k) = g(n - \lfloor n/k \rfloor - 1, k)$
```

Comment:
elaborate?

Reply:
They are patterns in those particular grundy values.  Idk how to demonstrate.

Reply:
Added suggestion for looking after patterns.